what-is-pair.org

```diff
@@ -0,0 +1,253 @@
+instead (haha, i think, this can definitely be corrected). ~pair-ssh-instance~
+is faster and more reliable than tmate though, and we could have it setup to
+start a tmux session instead that we attach to.
+
+In other words, there is an large amount of freedom given to people using
+environment. We offer heavily recommended suggestions, but few constraints.
+
+Also, there are flows that work great with Pair, but are not emphasised in its
+architecture. For example, running [[https://tilt.dev/][Tilt]] as you develop in cluster, so all
+changes are immediately deployed to the cluster and accessible via url. This is
+handy, and supported, but not advertised.
+
+** Our Emacs configuration
+Emacs is a major part of our Environment setup.  It can be broken into four layers:
+- Emacs the program
+- Doom Emacs, a framework for using and configuring Emacs
+- Our custom ii elisp packages, that extend emacs support for tmate, iterm, and
+  developing with postgres
+- Our custom doom configuration, which installs the above packages and a number
+  of others, in an idiosyncratic ii custom way.
+
+We also cache our current version of doom, as it can have a long install time and we
+want clusters to come up quickly.
+
+I bring up the layers as each need to be synced to make sure emacs works
+correctly. Emacs, doom emacs, our emacs packages, and our doom emacs config are
+each versioned, and if there is an update to one of them, without a coordinate
+update to the rest of them, then it is likely emacs will not work properly.
+Because of this, we've been conservative with updates to our custom
+configuration.
+
+The biggest difference in our configuration is the support of tmate blocks in org-mode,
+and to have these blocks open up appropriate windows in the Environment tmate session. This
+was developed so we could write executable documents in cluster, and see the output in our
+shared tmate session.
+
+* Comparisons to other solutions
+** Kind and Minikube
+Kind and Minikube are two ways to quickly run kubernetes clusters, and are often
+recommended for trying out kubernetes or trying out some new tool for kubernetes.
+
+You can compare them to Pair as all offer a simple way to create a cluster relatively quickly.
+
+The biggest differences are:
+- Pair is a production level cluster, running on bare metal equinix servers.  Kind and minikube
+  run inside docker and a virtual box respectively.  They cannot simulate a production-grade environment.
+- Pair can be easily setup as a multi-node cluster.  This is impossible with Minikube and difficult with Kind.
+- Pair offers built-in, simple DNS and name management. Neither come by default in Kind or Minikube. It is much
+  easier to access your Pair Environment and Pair cluster remotely.
+- Pair comes with an in-cluster coding environment.
+- Pair has a method for deploying services to yoru cluster given nothing but the repository link (assuming the repo has
+  a proper .sharing.io/init file)
```

Comment:
I think this could maybe use a bit more explaining?

Reply:
Agree, this part of pair saves me so much time every time [I use it](https://github.com/heyste/.sharing.io/blob/conformance/users/heyste/init)

Reply:
there is a dedicated section to this starting on line 63, under "services chosen by the project".  Should it be explained more than this?  I think this is a killer feature of Pair, but it feels straightforward: a shell script in an agreed-upon location that runs after the repo is cloned.

What makes an init file special is how the project owners write it, which is outside of Pair.   In the earlier section, though, I use APISnoop as an example of how the init file can be used to create a fully testing environment on launch.